Change cookie storage scope from replicator to database

pasin opened the pull request:
- Added cookieStorages properties to the internal CBLDatabase.
- Cookies are stored in the local database info document under the key named `cookies`.
- Implemented migrateOldCookieStorage to migrate cookies from checkpoint docs to the new local data info document. The old cookies are removed after the migration.
- CBLReplication always accesses the database cookie storage only from the background thread.

#1334

snej commented on Source/API/CBLReplication.m:
```diff
@@ -54,6 +53,11 @@ @interface CBLReplication ()
 
+@interface CBLReplication (Internal)
+// For unit test only:
+@property (nonatomic, readonly) NSArray* cookies;
```
Please move this to CouchbaseLitePrivate.h.

snej commented on Source/CBLDatabase+Internal.m:
```diff
@@ -813,4 +817,67 @@ - (void) postNotification: (NSNotification*)notification {
+    if (key == nil || value == nil)
+        return NO;
+    
+    NSMutableDictionary* document = [NSMutableDictionary dictionaryWithDictionary:
```
This will crash if the document doesn't already exist.

snej commented on Source/CBLBlipReplicator.m:
```diff
@@ -140,6 +139,12 @@ - (void) databaseClosing {
 @synthesize savingCheckpoint=_savingCheckpoint, active=_active;
 #endif
 
+- (CBLCookieStorage*) cookieStorage {
```
I don't think it's worth caching our own reference — it's easy enough to call `_db.cookieStorage` instead of `self.cookieStorage`.

snej commented on Source/CBLCookieStorage.m:
```diff
@@ -245,13 +248,39 @@ - (BOOL) deleteCookie: (NSHTTPCookie*)aCookie outIndex: (NSUInteger*)outIndex {
+
 - (void) loadCookies {
-    _cookies = [NSMutableArray array];
+    NSArray* cookies = [_db getLocalDatabaseInfoPropertyValueForKey: kLocalDbInfoCookiesKey];
```
Use `[_db.storage infoForKey:]` instead.

snej commented on Source/CBLDatabase+Internal.h:
```diff
@@ -153,4 +155,22 @@ extern NSArray* CBL_RunloopModes;
+// LOCAL DOC DATABASE INFO:
+
+/** Put a property with a given key and value into the local database info document. */
+- (BOOL) putLocalDatabaseInfoWithKey: (NSString*)key
```
We don't need these — can use CBL_Storage's "info" API.

snej commented on Source/CBLRestReplicator.m:
```diff
@@ -82,7 +82,7 @@ @implementation CBLRestReplicator
 }
 
-@synthesize db=_db, settings=_settings, cookieStorage=_cookieStorage, serverCert=_serverCert;
+@synthesize db=_db, settings=_settings, serverCert=_serverCert, cookieStorage=_cookieStorage;
```
I don't think we need `_cookieStorage` anymore. Can just access `_db.cookieStorage`.

snej commented on Unit-Tests/CookieStorage_Tests.m:
```diff
@@ -16,6 +18,12 @@ @interface CookieStorage_Tests : CBLTestCaseWithDB
 
+
+@interface CBLCookieStorage (Internal)
+- (NSString*) localDatabaseInfoCookiesKey;
```
Don't redeclare this; it should be imported from CouchbaseLitePrivate.h or CBLInternal.h.

pasin commented:
@snej Thanks for the code review. I made changes per your comment.

Note that I declare CBLCookieStorage `reset` method in `CBLInternal.h` instead of `CouchbaseLitePrivate.h` as CouchbaseLitePrivate.h doesn't seem to include any declaration for NON-public API interfaces.

snej commented on Source/CBLCookieStorage.m:
```diff
@@ -364,6 +402,17 @@ - (void) notifyCookiesChanged {
 }
 
+
+- (NSString*) databaseInfoCookiesKey {
```
I don't think this is used anywhere; can it be removed?

snej commented on Source/CBLDatabase+Internal.m:
```diff
@@ -54,6 +55,9 @@
 #define kHousekeepingDelayAfterOpening 3.0
 
+// Local document for keeping internal database info, e.g. cookies)
+#define kLocalDatabaseInfoDocId @"CBL_DatabaseInfo"
```
This doesn't seem to be used anywhere.

snej commented on Source/CBLDatabase+Replication.m:
```diff
@@ -108,6 +108,26 @@ - (BOOL) putLocalCheckpointDocumentWithKey: (NSString*)key
+    if (key == nil)
+        return NO;
+    
+    NSMutableDictionary* document = [NSMutableDictionary dictionaryWithDictionary:
```
This will crash if -getLocalCheckpointDocument returns nil. Instead you can use `[[[self getLocalCheckpointDocument] mutableCopy]`.

snej commented on Source/CBLInternal.h:
```diff
@@ -160,4 +161,9 @@ NSString* CBLKeyPathForQueryRow(NSString* keyPath); // for testing
+
+
+@interface CBLCookieStorage ()
+- (void) reset;
```
Why does this method need to be in a separate header? It seems like part of the class's regular API.

pasin replied:
Yes. Actually I can make changes to the existing deleteAllCookies: method to set nil to the database info cookies key.

pasin replied:
Actually I think it's better to rename deleteAllCookies: to reset:.

pasin commented:
@snej Thanks for the second code review. I removed unused variables methods are removed now and also fixed the potential crash. I changed the exiting deleteAllCookies to reset and included the logic to reset the database info cookies key as well. Please rereview when you have time.

pasin commented:
Rebase push as one fix didn't go through.

pasin commented:
Unit tests failed due to the change to the put / remove local checkpoint. I'm checking it.

pasin commented:
OK, I rebased the fixes. It's ready for re-review now. Thanks!

snej commented:
Great!

borrrden commented on Source/CBLCookieStorage.h:
```diff
@@ -25,7 +25,7 @@ extern NSString* const CBLCookieStorageCookiesChangedNotification;
 /** Creates a cookie storage that will store cookies inside the given database as a local document 
     referenced by a unique storage key. */
-- (instancetype) initWithDB: (CBLDatabase*)db storageKey: (NSString*)storageKey;
+- (instancetype) initWithDB: (CBLDatabase*)db;
```
This is a change to the public API!

snej replied:
CBLCookieStorage is not a public class.

borrrden replied:
Oops, my bad.  For some reason it is in .NET.  I forgot Obj-C classes are in the API folder.
